includes/embeds/class-amp-playlist-embed-handler.php

```diff
@@ -0,0 +1,307 @@
+<?php
+/**
+ * Class AMP_Playlist_Embed_Handler
+ *
+ * @package AMP
+ * @since 0.7
+ */
+
+/**
+ * Class AMP_Playlist_Embed_Handler
+ *
+ * Creates AMP-compatible markup for the WordPress 'playlist' shortcode.
+ *
+ * @package AMP
+ */
+class AMP_Playlist_Embed_Handler extends AMP_Base_Embed_Handler {
+
+	/**
+	 * The tag of the shortcode.
+	 *
+	 * @var string.
+	 */
+	const SHORTCODE = 'playlist';
+
+	/**
+	 * The max width of the audio thumbnail image.
+	 *
+	 * This corresponds to the max-width in wp-mediaelement.css:
+	 * .wp-playlist .wp-playlist-current-item img
+	 *
+	 * @var string.
+	 */
+	const THUMB_MAX_WIDTH = 60;
+
+	/**
+	 * The height of the carousel.
+	 *
+	 * @var string.
+	 */
+	const CAROUSEL_HEIGHT = 160;
+
+	/**
+	 * The pattern to get the playlist data.
+	 *
+	 * @var string.
+	 */
+	const PLAYLIST_REGEX = '/(?s)\<script [^>]* class="wp-playlist-script"\>[^<]*?(.*).*?\<\/script\>/';
+
+	/**
+	 * The ID of individual playlist.
+	 *
+	 * @var int
+	 */
+	public static $playlist_id = 0;
+
+	/**
+	 * The parsed data for the playlist.
+	 *
+	 * @var array
+	 */
+	public $data;
+
+	/**
+	 * Registers the playlist shortcode.
+	 */
+	public function register_embed() {
+		add_shortcode( self::SHORTCODE, array( $this, 'shortcode' ) );
+		add_action( 'wp_enqueue_scripts', array( $this, 'styling' ) );
+	}
+
+	/**
+	 * Unregisters the playlist shortcode.
+	 *
+	 * @return void.
+	 */
+	public function unregister_embed() {
+		remove_shortcode( self::SHORTCODE );
```

Comment:
I think the other examples of embeds in the plugin aren't quite right in how they call `remove_shortcode()`. This should actually be _restoring_ the previous shortcode that was added. So in the `register_embed` method instead of:

```php
add_shortcode( self::SHORTCODE, array( $this, 'shortcode' ) );
```

It should do:

```php
global $shortcode_tags;
if ( shortcode_exists( self::SHORTCODE ) ) {
    $this->removed_shortcode = $shortcode_tags[ self::SHORTCODE ];
}
add_shortcode( self::SHORTCODE, array( $this, 'shortcode' ) );
```

And then in `unregister_embed` it should do:

```
add_shortcode( self::SHORTCODE, $this->removed_shortcode );
```

Reply:
Thanks, [this commit](https://github.com/Automattic/amp-wp/pull/954/commits/d3cc38643431667544054fd1fa98bb5cc1a6c4c9) applies your suggestion of restoring the shortcode before `register_embed()` changed it.